jmclient/jmclient/taker.py

```diff
@@ -567,24 +567,48 @@ def on_sig(self, nick, sigb64):
                 break
             if len(sig_deserialized) == 2:
                 ver_sig, ver_pub = sig_deserialized
-                wit = None
+                scriptCode = None
             elif len(sig_deserialized) == 3:
-                ver_sig, ver_pub, wit =  sig_deserialized
+                ver_sig, ver_pub, scriptCode =  sig_deserialized
             else:
                 jlog.debug("Invalid signature message - more than 3 items")
                 break
-            ver_amt = utxo_data[i]['value'] if wit else None
+            ver_amt = utxo_data[i]['value'] if scriptCode else None
             sig_good = btc.verify_tx_input(txhex, u[0], utxo_data[i]['script'],
-                                               ver_sig, ver_pub, witness=wit,
-                                               amount=ver_amt)
+                            ver_sig, ver_pub, scriptCode=scriptCode, amount=ver_amt)
+
+            if ver_amt is not None and not sig_good:
+                # Special case to deal with legacy bots 0.5.0 or lower:
+                # the third field in the sigmessage was originally *not* the
+                # scriptCode, but the contents of tx['ins'][index]['script'],
+                # i.e. the witness program 0014... ; for this we can verify
+                # implicitly, as verify_tx_input used to, by reconstructing
+                # from the public key. For these cases, we can *assume* that
+                # the input is of type p2sh-p2wpkh
+                scriptCode = "76a914" + btc.hash160(unhexlify(ver_pub)) + "88ac"
```

Comment:
Initially I fully agreed with this, and wanted to delegate it to the wallet, but then realised: we can't do that, because the wallet may not know how to get the scriptCode (if e.g. it's a LegacyWallet, see e.g. the mixed maker test in test_coinjoin), it's fundamentally a general bitcoin operation; we *can* put it in cryptoengine but that comes back to the question of whether we really do want a layer on top of the jmbitcoin package there (see my earlier comments; I'm ambivalent about it). So currently the above has been replaced with a call to `btc.pubkey_to_p2pkh_script` (which is the same as you'd get by calling wallet.get_script_code if your wallet knew this particular script).

Reply:
I must admit I'm not 100% sure myself how this should ideally be handled. Adding wrappers to the wallet/cryptoengine to do wallet-independent operations does not seem like good design. I think delegating to jmbitcoin here is the correct thing to do.